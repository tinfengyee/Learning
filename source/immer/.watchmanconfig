{
  "ignore_dirs": ["node_modules", "_site", "dist", "coverage"]
}
