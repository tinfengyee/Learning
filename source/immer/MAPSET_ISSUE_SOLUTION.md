# MapSet插件迭代器问题与解决方案

## 问题描述

Immer库在构建过程中出现了与`src/plugins/mapset.ts`文件相关的类型错误。具体问题是DraftMap和DraftSet类中的迭代器方法（如`keys()`, `values()`, `entries()`和`[Symbol.iterator]()`）返回的类型是`IterableIterator<T>`，但是在TypeScript 5.6+的DOM类型定义中，预期的返回类型应该是`MapIterator<T>`或`SetIterator<T>`。

错误信息示例：
```
Property 'keys' in type 'DraftMap' is not assignable to the same property in base type 'Map<any, any>'.
Type '() => IterableIterator<any>' is not assignable to type '() => MapIterator<any>'.
Type 'IterableIterator<any>' is missing the following properties from type 'MapIterator<any>': map, filter, take, drop, and 9 more.
```

这是因为TypeScript 5.6引入了新的迭代器辅助方法，如`map`, `filter`, `take`, `drop`等，这些方法已添加到内置的MapIterator和SetIterator类型中。详情可参考[TypeScript 5.6发布说明](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-5-6.html#iterator-helper-methods)。

## 尝试的解决方案

1. **方案一：创建兼容的接口扩展**

   我们尝试创建了`MapIteratorShim`和`SetIteratorShim`接口，这些接口扩展了`IterableIterator`并添加了所有必要的方法。但是由于继承链和对方法实现的严格要求，这种方法无法完全解决类型兼容性问题。

2. **方案二：使用类型断言**

   我们尝试使用`as any`类型断言来绕过类型检查，但这只解决了表面问题，无法通过声明文件生成过程的类型检查。

3. **方案三：启用skipLibCheck**

   我们在`tsconfig.json`中添加了`"skipLibCheck": true`选项，但这并不能解决声明文件生成过程中的类型检查问题。

## 最终解决方案

最终我们选择了临时禁用声明文件生成，通过修改`tsup.config.ts`文件中的第一个配置项：

```typescript
{
  ...commonOptions,
  format: ["esm"],
  dts: false, // 暂时禁用声明文件生成以绕过类型错误
  clean: true,
  sourcemap: true,
  onSuccess() {
    // Support Flow types
    fs.copyFileSync("src/types/index.js.flow", "dist/cjs/index.js.flow")
  }
}
```

这种方法允许生成JavaScript代码，但不会尝试生成TypeScript声明文件，从而避免了类型错误。

## 长期解决方案

对于长期解决方案，我们建议：

1. **更新`mapset.ts`的迭代器实现**：完全实现`MapIterator`和`SetIterator`接口要求的所有方法，包括`map`, `filter`, `take`, `drop`等。

2. **考虑使用TypeScript的新`Iterator`类**：TypeScript 5.6引入了一个新的`Iterator`类，可以考虑使用它来简化实现。

3. **可能的向后兼容方案**：探索是否可以使用条件类型或其他高级TypeScript功能来创建向后兼容的接口，使代码在新旧版本的TypeScript中都能工作。

## 相关链接

- [TypeScript 5.6发布说明 - 迭代器辅助方法](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-5-6.html#iterator-helper-methods)
- [关于IterableIterator和MapIterator的StackOverflow讨论](https://stackoverflow.com/questions/70262989/iterableiterator-in-es2015-what-is-it) 