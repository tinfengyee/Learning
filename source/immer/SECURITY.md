# Security Policy

## Supported Versions

Latest only

## Reporting a Vulnerability

Security issues can be reported at info@michel.codes. Since this software is provided as-is no follow up, remediation or time lines are guaranteed.
