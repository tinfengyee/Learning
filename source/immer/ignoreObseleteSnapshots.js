module.exports = function(results) {
	// don't count obselete snapshot as a failure, but just check if there are no failing tests
	// console.dir(results)
	results.success = results.testResults.every(r => r.numFailingTests === 0)
	return results
}
