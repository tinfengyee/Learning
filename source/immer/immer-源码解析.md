# Immer 源码解析

## 1. 概述

Immer 是一个不可变数据结构库，它通过提供一种简单的 API 来实现不可变状态更新。Immer 的核心理念是让开发者能够以直接修改对象的方式来创建不可变数据的下一个状态。Immer 使用了 JavaScript 的 Proxy 特性来实现这一功能。

## 2. 源码结构

```
src/
├── core/                # 核心实现
│   ├── immerClass.ts    # Immer 主类
│   ├── proxy.ts         # Proxy 实现
│   ├── scope.ts         # 作用域管理
│   ├── finalize.ts      # 状态最终化处理
│   └── current.ts       # 获取当前状态
├── plugins/             # 插件系统
├── utils/               # 辅助函数
├── types/               # 类型定义
├── immer.ts             # 主入口
└── internal.ts          # 内部导出
```

## 3. 核心实现解析

### 3.1 `immer.ts` - 主入口文件

```typescript
// 创建一个 Immer 实例
const immer = new Immer()

/**
 * produce 函数接收一个初始状态和一个修改函数
 * 修改函数可以自由地修改传入的草稿状态
 * 所有修改只会应用到初始状态的副本上
 */
export const produce = immer.produce
```

主要职责：
- 创建 Immer 实例
- 导出核心 API 函数（produce, produceWithPatches 等）

### 3.2 `immerClass.ts` - Immer 类实现

```typescript
export class Immer implements ProducersFns {
  // 是否自动冻结对象
  autoFreeze_: boolean = true 
  
  // produce 函数实现
  produce: IProduce = (base: any, recipe?: any, patchListener?: any) => {
    // 处理柯里化调用
    if (typeof base === "function" && typeof recipe !== "function") {
      // ...
    }
    
    // 只有可草稿化的对象才会被处理
    if (isDraftable(base)) {
      const scope = enterScope(this)
      const proxy = createProxy(base, undefined)
      try {
        result = recipe(proxy)
        // ...
      } finally {
        // ...
      }
      return processResult(result, scope)
    } else {
      // 处理原始类型
      // ...
    }
  }
  
  // 带补丁的 produce 函数
  produceWithPatches: IProduceWithPatches = (base: any, recipe?: any): any => {
    // ...返回 [result, patches, inversePatches]
  }
  
  // 手动创建草稿
  createDraft<T extends Objectish>(base: T): Draft<T> {
    // 创建草稿但不立即处理
  }
  
  // 完成草稿处理
  finishDraft<D extends Draft<any>>(draft: D, patchListener?: PatchListener) {
    // 完成草稿处理并返回结果
  }
  
  // 应用补丁
  applyPatches<T extends Objectish>(base: T, patches: readonly Patch[]): T {
    // 应用补丁到对象上
  }
}
```

关键函数解释：
- `produce`: 核心函数，处理状态转换逻辑
- `produceWithPatches`: 类似 produce，但也返回补丁信息
- `createDraft`/`finishDraft`: 手动创建和完成草稿的函数
- `applyPatches`: 应用补丁到对象

### 3.3 `proxy.ts` - 代理实现

```typescript
// 创建代理对象
export function createProxyProxy<T extends Objectish>(
  base: T,
  parent?: ImmerState
): Drafted<T, ProxyState> {
  const isArray = Array.isArray(base)
  const state: ProxyState = {
    // 对象类型（数组或对象）
    type_: isArray ? ArchType.Array : ArchType.Object,
    // 当前作用域
    scope_: parent ? parent.scope_ : getCurrentScope()!,
    // 是否被修改
    modified_: false,
    // 是否已完成
    finalized_: false,
    // 跟踪哪些属性被赋值或删除
    assigned_: {},
    // 父状态
    parent_: parent,
    // 基础状态（原始对象）
    base_: base,
    // 代理对象（将设置）
    draft_: null as any,
    // 对象的副本（延迟创建）
    copy_: null,
    // 撤销函数
    revoke_: null as any,
    // 是否为手动模式
    isManual_: false
  }
  
  // 创建代理
  const {revoke, proxy} = Proxy.revocable(
    isArray ? [state] as any : state as any,
    isArray ? arrayTraps : objectTraps
  )
  state.draft_ = proxy as any
  state.revoke_ = revoke
  return proxy as any
}

// 对象代理拦截器
export const objectTraps: ProxyHandler<ProxyState> = {
  // 获取属性
  get(state, prop) {
    // 处理特殊属性和代理状态获取
    if (prop === DRAFT_STATE) return state
    
    const source = latest(state)
    const value = source[prop]
    
    // 如果值可以被草稿化且尚未被修改
    if (!state.finalized_ && isDraftable(value)) {
      // 检查是否需要创建子代理
      if (value === peek(state.base_, prop)) {
        prepareCopy(state)
        return (state.copy_![prop] = createProxy(value, state))
      }
    }
    return value
  },
  
  // 设置属性
  set(state, prop, value) {
    // 处理 setter
    const desc = getDescriptorFromProto(latest(state), prop)
    if (desc?.set) {
      desc.set.call(state.draft_, value)
      return true
    }
    
    // 检查是否需要标记为修改
    if (!state.modified_) {
      const current = peek(latest(state), prop)
      // 如果值相同，无需修改
      if (is(value, current) && (value !== undefined || has(state.base_, prop)))
        return true
      
      // 准备副本并标记为已修改
      prepareCopy(state)
      markChanged(state)
    }
    
    // 设置新值
    state.copy_![prop] = value
    state.assigned_[prop] = true
    return true
  },
  
  // 删除属性
  deleteProperty(state, prop) {
    // 标记为已删除
    if (peek(state.base_, prop) !== undefined || prop in state.base_) {
      state.assigned_[prop] = false
      prepareCopy(state)
      markChanged(state)
    }
    if (state.copy_) {
      delete state.copy_[prop]
    }
    return true
  },
  
  // 其他拦截器...
}
```

关键函数解释：
- `createProxyProxy`: 创建代理对象，跟踪对象状态
- `objectTraps`: 对象代理拦截器，处理属性读取/写入/删除
- `arrayTraps`: 数组代理拦截器，特殊处理数组操作
- `markChanged`: 标记对象已修改
- `prepareCopy`: 准备对象副本（延迟复制优化）

### 3.4 `finalize.ts` - 最终化处理

```typescript
// 处理 produce 函数的结果
export function processResult(result: any, scope: ImmerScope) {
  scope.unfinalizedDrafts_ = scope.drafts_.length
  const baseDraft = scope.drafts_![0]
  const isReplaced = result !== undefined && result !== baseDraft
  
  if (isReplaced) {
    // 处理返回新值的情况
    if (isDraftable(result)) {
      result = finalize(scope, result)
      // 可能冻结结果
      if (!scope.parent_) maybeFreeze(scope, result)
    }
    
    // 生成替换补丁
    if (scope.patches_) {
      getPlugin("Patches").generateReplacementPatches_(/*...*/)
    }
  } else {
    // 处理修改草稿的情况
    result = finalize(scope, baseDraft, [])
  }
  
  // 清理和补丁处理
  revokeScope(scope)
  if (scope.patches_) {
    scope.patchListener_!(scope.patches_, scope.inversePatches_!)
  }
  
  return result !== NOTHING ? result : undefined
}

// 将草稿最终化为不可变对象
function finalize(rootScope: ImmerScope, value: any, path?: PatchPath) {
  // 已冻结对象直接返回
  if (isFrozen(value)) return value
  
  const state: ImmerState = value[DRAFT_STATE]
  
  // 处理普通对象（非草稿）
  if (!state) {
    each(value, (key, childValue) =>
      finalizeProperty(rootScope, state, value, key, childValue, path)
    )
    return value
  }
  
  // 不处理其他作用域的草稿
  if (state.scope_ !== rootScope) return value
  
  // 未修改的草稿，返回（可能被冻结的）原始值
  if (!state.modified_) {
    maybeFreeze(rootScope, state.base_, true)
    return state.base_
  }
  
  // 最终化已修改但未完成的草稿
  if (!state.finalized_) {
    state.finalized_ = true
    state.scope_.unfinalizedDrafts_--
    
    // 处理所有子属性的最终化
    // ...
    
    // 可能冻结结果
    maybeFreeze(rootScope, state.copy_, false)
    
    // 生成补丁
    if (path && rootScope.patches_) {
      getPlugin("Patches").generatePatches_(/*...*/)
    }
  }
  
  return state.copy_
}
```

关键函数解释：
- `processResult`: 处理 produce 函数的结果
- `finalize`: 将草稿转换为最终的不可变状态
- `finalizeProperty`: 处理单个属性的最终化
- `maybeFreeze`: 在适当条件下冻结对象

### 3.5 `scope.ts` - 作用域管理

```typescript
/** 每个作用域代表一次 produce 调用 */
export interface ImmerScope {
  // 补丁和反向补丁
  patches_?: Patch[]
  inversePatches_?: Patch[]
  // 是否可以自动冻结
  canAutoFreeze_: boolean
  // 当前作用域的草稿列表
  drafts_: any[]
  // 父作用域
  parent_?: ImmerScope
  // 补丁监听器
  patchListener_?: PatchListener
  // Immer 实例
  immer_: Immer
  // 未完成的草稿数量
  unfinalizedDrafts_: number
}

// 进入一个新的作用域
export function enterScope(immer: Immer) {
  return (currentScope = createScope(currentScope, immer))
}

// 离开当前作用域
export function leaveScope(scope: ImmerScope) {
  if (scope === currentScope) {
    currentScope = scope.parent_
  }
}

// 撤销作用域
export function revokeScope(scope: ImmerScope) {
  leaveScope(scope)
  scope.drafts_.forEach(revokeDraft)
  scope.drafts_ = null as any
}
```

关键函数解释：
- `enterScope`: 进入新的作用域，创建新的上下文
- `leaveScope`: 离开当前作用域，恢复父作用域
- `revokeScope`: 撤销作用域，释放所有草稿
- `usePatchesInScope`: 在作用域中启用补丁跟踪

### 3.6 `common.ts` - 通用工具函数

```typescript
/** 判断值是否为草稿 */
export function isDraft(value: any): boolean {
  return !!value && !!value[DRAFT_STATE]
}

/** 判断值是否可以被草稿化 */
export function isDraftable(value: any): boolean {
  if (!value) return false
  return (
    isPlainObject(value) ||
    Array.isArray(value) ||
    !!value[DRAFTABLE] ||
    !!value.constructor?.[DRAFTABLE] ||
    isMap(value) ||
    isSet(value)
  )
}

/** 获取草稿的原始对象 */
export function original(value: Drafted<any>): any {
  if (!isDraft(value)) die(15, value)
  return value[DRAFT_STATE].base_
}

/** 获取当前的对象状态（副本或原始） */
export function latest(state: ImmerState): any {
  return state.copy_ || state.base_
}

/** 浅拷贝对象 */
export function shallowCopy(base: any, strict: StrictMode) {
  // 根据不同类型创建浅拷贝
  if (isMap(base)) return new Map(base)
  if (isSet(base)) return new Set(base)
  if (Array.isArray(base)) return Array.prototype.slice.call(base)
  
  // 处理普通对象
  // ...
}

/** 冻结对象 */
export function freeze<T>(obj: any, deep: boolean = false): T {
  if (isFrozen(obj) || isDraft(obj) || !isDraftable(obj)) return obj
  // 冻结不同类型的对象
  // ...
}
```

关键函数解释：
- `isDraft`: 检查对象是否为草稿
- `isDraftable`: 检查对象是否可被草稿化
- `original`: 获取草稿的原始对象
- `latest`: 获取最新状态（副本或原始）
- `shallowCopy`: 创建对象的浅拷贝
- `freeze`: 冻结对象，可选深度冻结

## 4. 核心流程

Immer 的工作流程可以总结为以下几个步骤：

1. **初始化**：创建 Immer 实例
2. **调用 produce**：传入原始状态和修改函数
3. **创建代理**：为原始状态创建代理对象（草稿）
4. **执行用户函数**：执行修改函数，对草稿进行修改
5. **跟踪变更**：通过 Proxy 拦截器跟踪所有修改
6. **最终化**：将草稿转换为最终的不可变状态
   - 未修改的属性直接引用原始状态
   - 已修改的属性创建新副本
7. **生成补丁**（可选）：如果指定了补丁监听器，生成变更补丁
8. **返回结果**：返回新的不可变状态

## 5. 性能优化

Immer 采用了多种优化措施：

1. **结构共享**：未修改的部分直接引用原始状态，避免不必要的复制
2. **延迟复制**：只有在属性被修改时才创建副本（copy-on-write）
3. **缓存检测**：使用 `modified_` 标志避免重复检查

## 6. 使用方法

### 6.1 基本用法

```javascript
import produce from "immer"

const baseState = {
  users: [{ name: "张三" }],
  settings: { theme: "dark" }
}

const nextState = produce(baseState, draft => {
  draft.users.push({ name: "李四" })
  draft.settings.theme = "light"
})
```

### 6.2 柯里化用法

```javascript
const updateUser = produce((draft, name) => {
  draft.users.push({ name })
})

const nextState = updateUser(baseState, "王五")
```

### 6.3 补丁和撤销

```javascript
import { produceWithPatches, applyPatches } from "immer"

const [nextState, patches, inversePatches] = produceWithPatches(
  baseState,
  draft => {
    draft.users.push({ name: "赵六" })
  }
)

// 应用补丁
const patchedState = applyPatches(baseState, patches)

// 撤销操作（应用反向补丁）
const revertedState = applyPatches(nextState, inversePatches)
```

## 7. 总结

Immer 的核心优势在于：

1. **简化复杂嵌套状态更新**：提供直接修改的编程模型
2. **保持不可变性**：内部自动处理不可变转换
3. **结构共享**：优化性能和内存使用
4. **补丁功能**：支持变更跟踪和时间旅行功能

Immer 通过巧妙使用 JavaScript Proxy，实现了一种既易于使用又保持不可变性的状态管理方案，是现代前端状态管理的优秀工具。 