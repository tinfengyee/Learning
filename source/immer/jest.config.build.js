module.exports = {
	moduleNameMapper: {
		"src/.*": "<rootDir>/dist/cjs/immer.cjs.production.js"
	},
	testEnvironmentOptions: {
		url: "http://localhost"
	},
	globals: {
		USES_BUILD: true
	},
	preset: "ts-jest/presets/js-with-ts",
	testEnvironment: "node",
	testMatch: ["**/__tests__/**/*.[jt]s?(x)"],
	snapshotResolver: "<rootDir>/jest.config.build.snapshots.js",
	testResultsProcessor: "<rootDir>/ignoreObseleteSnapshots.js"
}
