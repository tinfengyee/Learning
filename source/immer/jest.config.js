module.exports = {
	testEnvironmentOptions: {
		url: "http://localhost"
	},
	preset: "ts-jest/presets/js-with-ts",
	testEnvironment: "node",
	testMatch: ["**/__tests__/**/*.[jt]s?(x)"]
}
