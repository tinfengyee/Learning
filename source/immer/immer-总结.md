# Immer 库总结

## 核心原理

Immer 是一个用于简化不可变数据结构操作的 JavaScript 库，它基于以下核心原理：

1. **代理拦截**：使用 JavaScript 的 `Proxy` 对象拦截对数据的所有操作。
2. **草稿状态**：提供一个可修改的"草稿"，用户可以直接修改它。
3. **结构共享**：只复制修改的部分，未修改的部分保持对原始对象的引用。
4. **自动冻结**：最终返回冻结（不可变）的对象。

## 实现流程

Immer 的工作流程可以概括为以下步骤：

1. **创建代理**：为原始状态创建一个可修改的代理对象（草稿）。
2. **记录修改**：通过 Proxy 拦截器记录对草稿的所有修改操作。
3. **延迟复制**：只有在属性被修改时才创建该部分的副本（copy-on-write 模式）。
4. **最终处理**：根据记录的修改，创建一个新的不可变状态。
5. **结构共享**：未修改的部分直接引用原始状态。

## 关键实现点

1. **代理处理**：
   - 使用 `Proxy.revocable` 创建可撤销的代理
   - 针对对象和数组分别定义不同的拦截器

2. **状态追踪**：
   - 每个代理对象关联一个状态对象，记录修改信息
   - 使用 `modified_` 和 `assigned_` 标志跟踪变更

3. **作用域管理**：
   - 每次 `produce` 调用创建一个作用域
   - 支持嵌套的 `produce` 调用

4. **最终处理**：
   - 将草稿状态转换为最终的不可变状态
   - 处理嵌套对象和数组
   - 保持结构共享

5. **补丁系统**：
   - 生成描述变更的补丁（patches）
   - 支持撤销操作（通过逆补丁）

## 性能优化

Immer 采用了多种性能优化技术：

1. **延迟复制**：只有在实际修改时才创建副本
2. **结构共享**：最大限度减少内存使用
3. **缓存检测**：避免重复检查已知状态
4. **优化的相等性检查**：精细判断值是否实际改变

## 主要优势

1. **简化代码**：允许以直接修改的方式编写不可变更新逻辑。
2. **减少错误**：避免手动创建深层嵌套对象更新时的常见错误。
3. **提高可读性**：代码更接近实际的业务逻辑，减少样板代码。
4. **高性能**：通过结构共享和延迟复制优化性能。
5. **不可变保证**：确保返回的对象是真正不可变的。
6. **时间旅行能力**：通过补丁系统支持撤销/重做。

## 应用场景

1. **React 状态管理**：结合 React 的 `useState` 或 Redux 使用。
2. **复杂表单处理**：管理复杂表单的状态变化。
3. **撤销/重做功能**：利用补丁系统实现历史操作。
4. **任何需要不可变数据的场景**：如函数式编程、状态同步等。

## 使用示例

### 基本用法
```javascript
import produce from "immer";

const baseState = { 
  users: [{ name: "张三" }],
  settings: { theme: "dark" }
};

const nextState = produce(baseState, draft => {
  draft.users.push({ name: "李四" });
  draft.settings.theme = "light";
});
```

### 柯里化用法
```javascript
const toggleTheme = produce(draft => {
  draft.settings.theme = draft.settings.theme === "light" ? "dark" : "light";
});

const nextState = toggleTheme(baseState);
```

### 补丁和撤销
```javascript
import { produceWithPatches, applyPatches } from "immer";

const [nextState, patches, inversePatches] = produceWithPatches(
  baseState,
  draft => {
    draft.users.push({ name: "赵六" });
  }
);

// 撤销操作
const revertedState = applyPatches(nextState, inversePatches);
```

## 总结

Immer 通过巧妙利用 JavaScript 的 Proxy 特性，成功地将不可变数据的操作简化为类似可变数据的编程模型，同时保持了不可变性的所有优势。它是现代前端状态管理的强大工具，能够显著提高开发效率和代码质量。 