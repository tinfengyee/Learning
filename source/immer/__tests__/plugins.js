import {produce, produceWithPatches, applyPatches} from "../src/immer"

test("error when using Maps", () => {
	expect(() => {
		produce(new Map(), function() {})
	}).toThrowErrorMatchingSnapshot()
})

test("error when using patches - 1", () => {
	expect(() => {
		produce(
			{},
			function() {},
			function() {}
		)
	}).toThrowErrorMatchingSnapshot()
})

test("error when using patches - 2", () => {
	expect(() => {
		produceWithPatches({}, function() {})
	}).toThrowErrorMatchingSnapshot()
})

test("error when using patches - 3", () => {
	expect(() => {
		applyPatches({}, [])
	}).toThrowErrorMatchingSnapshot()
})
