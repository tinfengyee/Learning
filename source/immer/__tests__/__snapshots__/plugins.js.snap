// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`error when using Maps 1`] = `"[Immer] The plugin for 'MapSet' has not been loaded into Immer. To enable the plugin, import and call \`enableMapSet()\` when initializing your application."`;

exports[`error when using patches - 1 1`] = `"[Immer] The plugin for 'Patches' has not been loaded into Immer. To enable the plugin, import and call \`enablePatches()\` when initializing your application."`;

exports[`error when using patches - 2 1`] = `"[Immer] The plugin for 'Patches' has not been loaded into Immer. To enable the plugin, import and call \`enablePatches()\` when initializing your application."`;

exports[`error when using patches - 3 1`] = `"[Immer] The plugin for 'Patches' has not been loaded into Immer. To enable the plugin, import and call \`enablePatches()\` when initializing your application."`;
